### ISSUE TITLE
Overriding `index` action with async variant throws error

### ISSUE THREAD

**Issue by nickschot** [postlight/lux#603]
When overriding the index action in a controller (probably valid for the other by default non-async actions too) with an async variant `async index(req, res) { ... }` the following error is thrown:

```json
{
      "status": "500",
      "title": "Internal Server Error",
      "detail": "Cannot read property 'scopes' of undefined"
}
```

My use case for overriding the index action is that I have to do some more complex things which are not possible with the Lux ORM when certain query parameters are present which involves asynchronously retrieving data from the database after which I will retrieve the correct model instances via `super().where()` or via `MyModel.where()`.

The relevant stacktrace:
```bash
TypeError: Cannot read property 'scopes' of undefined
    at scopesFor (/node_modules/lux-framework/src/packages/database/query/utils/scopes-for.js:7:22)
    at new Query (/node_modules/lux-framework/src/packages/database/query/index.js:98:35)
    at Function.from (/node_modules/lux-framework/src/packages/database/query/index.js:444:26)
    at /node_modules/lux-framework/src/packages/router/route/action/enhancers/resource.js:22:15
    at Generator.next (<anonymous>)
    at step (/dist/bundle.js:77:30)
    at /dist/bundle.js:95:14
    at /dist/bundle.js:74:12
    at bound index (/node_modules/lux-framework/src/packages/router/route/action/enhancers/resource.js:12:9)
    at __FINAL_HANDLER__ (/node_modules/lux-framework/src/packages/router/route/action/index.js:26:7)
    at /node_modules/lux-framework/src/packages/router/route/action/enhancers/track-perf.js:15:20
    at Generator.next (<anonymous>)
    at step (/dist/bundle.js:77:30)
    at /dist/bundle.js:95:14
    at /dist/bundle.js:74:12
    at __FINAL_HANDLER__ (/node_modules/lux-framework/src/packages/router/route/action/enhancers/track-perf.js:12:9)
```

**Comment by nickschot** [comment]
This also happens when `index` is not async but when a promise is returned within index (when some async function is called within index and the result is returned from within that function).

**Comment by nickschot** [comment]
Extra test case from Gitter (controller body):

```javascript
index(req, res){
      return this.myFunction();
}

async myFunction(){
     let ids = await someCustomDbCallReturningAnArrayOfIds();
     return MyModel.where({id: ids})
}
```

**Comment by nickschot** [comment]
I've implemented a similar thing in a `create` action. Oddly enough, that works fine. When doing anything to do with promises (including normal promises), the scopes error occurs, so it likely means it's not (just?) caused by the use of async/await/generators.

Another test case (with a normal `.then`) which works in the create action but not in the index action (with the respective super call of course):
```javascript
create(request, response) {
    return this.controllers.get('application').db.from('some_table').max('id').then((maxId) => {
        return super.create(request, response);
    });
}
```

**Comment by nickschot** [comment]
The error doesn't seem to be directly caused by the code in the override as the models are correctly set for all calls to super or the model directly and the queries show correctly in the log.

The error originates from `lux-framework/src/packages/router/route/action/enhancers/resource.js` L19 and causes the 500 while (sometimes before, sometimes after) the overriding index method is called. Model is `undefined` in `scopes-for` for this call.

**Comment by nickschot** [comment]
Got it, it's caused by the `total` query on [L21](https://github.com/postlight/lux/blob/9e2f84bd2d0675023b694a154b01f0b28604206e/src/packages/router/route/action/enhancers/resource.js#L21) `Query.from(data).count()`. This does not resolve when index is overridden with some index function returning a promise.

cc @zacharygolba

**Comment by nickschot** [comment]
The promise returned from `result` also contains an error in the `_trace` key:

```sh
{ Error
       at Promise.longStackTracesCaptureStackTrace [as _captureStackTrace] (/node_modules/bluebird/js/release/debuggability.js:397:19)
       at Promise._then (/node_modules/bluebird/js/release/promise.js:231:17)
       at Promise.then (/node_modules/bluebird/js/release/promise.js:124:17)
       at AssignmentsController.index (/app/controllers/assignments.js:86:18)
       at /node_modules/lux-framework/src/packages/router/route/action/enhancers/resource.js:14:20
       at Generator.next (<anonymous>)
       at step (/dist/bundle.js:76:30)
       at /dist/bundle.js:94:14
       at /dist/bundle.js:73:12
       at bound index (/node_modules/lux-framework/src/packages/router/route/action/enhancers/resource.js:12:9)
       at __FINAL_HANDLER__ (/node_modules/lux-framework/src/packages/router/route/action/index.js:28:7)
       at /callsheet-lux/node_modules/lux-framework/src/packages/router/route/action/enhancers/track-perf.js:15:20
       at Generator.next (<anonymous>)
       at step (/dist/bundle.js:76:30)
       at /dist/bundle.js:94:14
       at /dist/bundle.js:73:12
       at __FINAL_HANDLER__ (/node_modules/lux-framework/src/packages/router/route/action/enhancers/track-perf.js:12:9)
       at /node_modules/lux-framework/src/packages/router/route/index.js:135:14
       at Generator.next (<anonymous>)
       at step (/dist/bundle.js:76:30)
       at /dist/bundle.js:94:14
       at /dist/bundle.js:73:12 _parent: undefined, _promisesCreated: 0, _length: 1 }
```

**Comment by zacharygolba** [comment]
More information on the reason/fix for this issue can be found in [this gitter conversaton](https://gitter.im/postlight/lux?at=5880c6d1e836bf70106465c3).

**Comment by zacharygolba** [comment]
Closed via #603
